#167392098 A user can view all available properties

Genza999 opened the pull request:
## Description ##
- Enable user to retrieve all available properties

## Type of change ##
Please select the relevant option
- [ ] Bug fix(a non-breaking change which fixes an issue)
- [x] New feature(a non-breaking change which adds functionality)
- [ ] Breaking change(fix of a feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

## How Has This Been Tested? ##
Please describe the tests that you ran to verify your changes
- [ ] End to end test
- [ ] Integration test
- [x] unit test

## Checklist: ##
- [x] My code follows the style guidelines of this project
- [x] I have linted my code prior to submission
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing tests pass locally with my changes

## Related Pivotal Tracker stories ##
[#167392098](https://www.pivotaltracker.com/story/show/167392098)

Commit:
ft(retreive-properties): user can retreive all available properties

- create service to retrieve properties from backend api
- create intrfaces for data types of retrieved properties
- design front-end layout for displaying properties retrieved

[Finishes #167392098]

Commit:
ft(retreive-properties): users can retrieve available properties

- include tests for services and component

[Finishes #167392098]

kelvinndmo commented on src/app/pages/properties/properties.component.ts:
```diff
@@ -0,0 +1,47 @@
+  toggle: boolean = true;
+
+  constructor(private propertiesServices: PropertiesService) {}
+
```
I thought the team agreed to use Reactive Forms instead of Template Driven Forms

Genza999 replied:
Hey @kelvinndmo , could you explain a bit more on this because I am not using any forms in this feature.

kelvinndmo replied:
I didn't realize that. Thanks for the clarification

MandelaK commented on src/app/layouts/common-layout/common-layout.module.ts:
```diff
@@ -1,20 +1,20 @@
 
-import {CommonLayoutRoutes} from './common-layout.routing';
-import {HomeComponent} from '../../pages/home/home.component';
+import { CommonLayoutRoutes } from "./common-layout.routing";
```
Please use absolute imports ..

Genza999 replied:
Hey @MandelaK , this is not part of my feature.

CryceTruly commented on src/app/models/Address.ts:
```diff
@@ -0,0 +1,5 @@
+export interface Address {
+  Street: string;
```
Good work @Genza999 , however, we should use camelCase for variable naming, its conventional in JS

Genza999 replied:
Thanks @CryceTruly , worked on this

CryceTruly commented on src/app/pages/properties/properties.component.ts:
```diff
@@ -0,0 +1,47 @@
+})
+export class PropertiesComponent implements OnInit {
+  properties: Property[];
+  propertiesUrl = "http://127.0.0.1:8000/api/v1/properties/";
```
Hello @Genza999, could you use the live Heroku url such that it's easier to test the functionality on the review app also.

Genza999 replied:
Thanks @CryceTruly, I have worked on this

Commit:
ft(retrieve properties): user can retrieve properties

- add on responsiveness of the layout

[Finishes #167392098]

Commit:
ft(retrieve-properties): user can retrieve properties

- create functionality to fetch token from local storage

[Finishes #167392098]

MandelaK commented on src/app/pages/properties/properties.component.ts:
```diff
@@ -0,0 +1,47 @@
+    this.setProperties(this.propertiesUrl);
+  }
+
+  setProperties(url: string) {
```
I noticed that when there are no property to display, the site almost looks broken ... Maybe you could add some information notifying the users that there are no property to display and they could check back later?

Genza999 replied:
Thanks @MandelaK , I have worked on this.

CryceTruly replied:
Hey @Genza999, on the same note when there are no properties, The footer is way in the middle of the page which does not look great,we could try and make it stick to the bottom.

Genza999 replied:
Hey @CryceTruly Thanks I have worked on this issue.

MandelaK reviewed (commented):
Everything looks good, however when viewing the page and there are no property, the page looks broken and there isn't any feedback to indicate that there are no property to display ... Do you think you could improve that?

Commit:
ft(retrieve-properties): user can retrieve all properties

- format the UI layout

[Finishes #167392098]

Commit:
ft(retrieve-properties): users can retrieve all properties

- add functionlity to show different content when no properties exist

[Finishes #167392098]

CryceTruly commented on package.json:
```diff
@@ -25,9 +25,11 @@
     "@angular/platform-browser": "~8.1.1",
     "@angular/platform-browser-dynamic": "~8.1.1",
     "@angular/router": "~8.1.1",
+    "angular-material": "^1.1.19",
```
@Genza999, did we use this library?

Genza999 replied:
Thanks @CryceTruly , I have removed it.

CryceTruly commented on src/app/models/Client.ts:
```diff
@@ -0,0 +1,8 @@
+import { Address } from "./Address";
+
+export interface Client {
+  client_name: string;
```
@Genza999, please use CameCalse here.

Genza999 replied:
I have changed it thanks to CamelCase

CryceTruly commented on src/app/models/Property.ts:
```diff
@@ -0,0 +1,42 @@
+export interface Property {
+  id?: number;
+  price?: number;
+  lost_size?: number;
```
@Genza999, did you mean to write lotSize?

Genza999 replied:
Thanks @CryceTruly, I have changed the name

CryceTruly commented on src/app/models/Property.ts:
```diff
@@ -0,0 +1,42 @@
+  image_main?: string;
+  video?: string;
+  view_count?: number;
+  last_viewed?: string;
```
@Genza999, please use cameCase for variable naming, Its customary JS convention.

Genza999 replied:
Hey @CryceTruly I can't because the structure of the data from the backend is in this format. Kindly elaborate on this since these are interfaces.

Genza999 replied:
Hey @CryceTruly I have changed it tp CamelCase thanks

CryceTruly commented on src/app/pages/properties/properties.component.spec.ts:
```diff
@@ -0,0 +1,179 @@
+    expect(component).toBeTruthy();
+  });
+
+  it("test getProperties function", () => {
```
Good work @Genza999, It's customary for `it` to be followed by `should`, We could we write our descriptions to follow the conventions and make them more descriptive.

Genza999 replied:
Thanks @CryceTruly I have worked on that

CryceTruly commented on src/index.html:
```diff
@@ -1,19 +1,26 @@
+    <link href="favicon.ico" rel="icon" type="image/x-icon" />
+    <!-- FontAwesome 4.7.0 -->
+    <link
+      href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/4.7.0/css/font-awesome.min.css"
```
Good work @Genza999, please remember to update the CDN link to use font awesome 5.
It provides more icons and is backward compatible.

Genza999 replied:
Thanks @CryceTruly i have worked on that.

Commit:
ft(retrieve-properties): user can retrieve all available properties

- change to camel case in the interfaces
- add landville image to properties page
- refactor tests decsription text

[Finishes #167392098]

Jaman-dedy commented:
Great job @Genza999, all the functionality is well implemented, I just wanted to suggest you some little changes to set the UX more friendly.
1. When the user is on the properties page, it can be better when he hovers a property to make a little change on the hovered property ex: `change a little bit the color, or the cursor can be a pointer, or a little transformation` 
2. About the pagination: the font size of the 2 buttons is too small, and they are not centered. If it can be possible that is visible. ex : 
![image](https://user-images.githubusercontent.com/46047244/62544251-b9325c00-b85f-11e9-99d7-5bda0e484eae.png)

Commit:
ft(retreive-properties): user can retrieve properties

- add loader to the properties component
- minify landville image
- add disabling of pagination buttons if no property exists
- fix properties link on home page
- add change of card color on hover action

[Finishes #167392098]

Genza999 commented:
> Great job @Genza999, all the functionality is well implemented, I just wanted to suggest you some little changes to set the UX more friendly.
> 
> 1. When the user is on the properties page, it can be better when he hovers a property to make a little change on the hovered property ex: `change a little bit the color, or the cursor can be a pointer, or a little transformation`
> 2. About the pagination: the font size of the 2 buttons is too small, and they are not centered. If it can be possible that is visible. ex :
>    ![image](https://user-images.githubusercontent.com/46047244/62544251-b9325c00-b85f-11e9-99d7-5bda0e484eae.png)



> Great job @Genza999, all the functionality is well implemented, I just wanted to suggest you some little changes to set the UX more friendly.
> 
> 1. When the user is on the properties page, it can be better when he hovers a property to make a little change on the hovered property ex: `change a little bit the color, or the cursor can be a pointer, or a little transformation`
> 2. About the pagination: the font size of the 2 buttons is too small, and they are not centered. If it can be possible that is visible. ex :
>    ![image](https://user-images.githubusercontent.com/46047244/62544251-b9325c00-b85f-11e9-99d7-5bda0e484eae.png)

Hey @Jaman-dedy , I have worked on these thanks

bl4ck4ndbr0wn commented:
![image](https://user-images.githubusercontent.com/33542127/62607868-77f58700-b907-11e9-8ba8-558864c0da1b.png)
figure1: Spacing, between components, is too large. try reducing it.

![image](https://user-images.githubusercontent.com/33542127/62607961-a3787180-b907-11e9-8ab6-2695da2a7731.png)

figure2: Pagination, try spacing it and move it to the left.

![image](https://user-images.githubusercontent.com/33542127/62608781-3239be00-b909-11e9-9e78-042f964ff372.png)
figure 3: Try to stretch the properties to reach the end of the column.

patrickf949 commented:
@Genza999 please work on your coverage

Commit:
ft(retrieve-users): users can retreive all available properties

- change button colors on click
- show pagination number of properties displayed
- align pagination buttons

[Finishes #167392098]

bl4ck4ndbr0wn reviewed (approved):
Great job on implementing feedback given.
LGTM

patrickf949 reviewed (approved):
LGTM

Victoradukwu reviewed (approved):
Nice Job!

CryceTruly reviewed (approved):
Thanks for the great Job done @Genza999

malep2007 commented on src/app/layouts/common-layout/common-layout.routing.ts:
```diff
@@ -1,12 +1,22 @@
-  { path: 'registersuccess', component: RegistersuccessComponent },
-  { path: 'profile', component: ProfileComponent }
+  {
+    path: "create-company",
```
I think we should adopt better conventions for URLs in the application. `create-company` is not the best way to have a URL on the browser, something like `company/create/` or `company/profile/` or even `company/register/` is more appropriate and better for users to recal.

malep2007 replied:
@Genza999 the url is still not conventional. Could you look into this? Also, since we are working on the frontend, you can read this [article](https://www.clickz.com/importance-of-url-naming-convention-and-url-management-for-websites/37490/)

Genza999 replied:
Hey @malep2007 , thankyou for the feedback but this was not part of my feature

JamesMudidi reviewed (approved):
LGTM

malep2007 commented on src/app/pages/profile/personal-information/personal-information.component.ts:
```diff
@@ -30,28 +30,28 @@ export class PersonalInformationComponent implements OnInit, OnDestroy {
-  phone = new FormControl('', [Validators.required, Validators.minLength(14)]);
-  employer = new FormControl('');
-  designation = new FormControl('');
+  street = new FormControl("", Validators.required);
```
Could you refactor your code to take advantage of the FormGroup() provided by Angular? If you need a sync on this let me know.

Genza999 replied:
Hey @malep2007 , thankyou for the feedback but this was not part of my feature. I think when I configured my linter, it then reflects that I changed most parts of the code

malep2007 commented on src/app/pages/properties/no-properties/no-properties.component.html:
```diff
@@ -0,0 +1,3 @@
+<p class="center">
+    Sorry, No properties available now. Kindly Come back later
+  </p>
```
@Genza999 new line.

Genza999 replied:
Thanks @malep2007 I have added a new line here

malep2007 commented on src/app/pages/properties/properties.component.ts:
```diff
@@ -0,0 +1,84 @@
+      } else {
+        this.properties = this.results;
+
+        if (response.data.properties.next) {
```
Consider using if else if for this section
```javascript
        if (response.data.properties.next) {
          this.next = response.data.properties.next;
        } else {
          this.disabledNext = true;
        }

        if (response.data.properties.previous) {
          this.previous = response.data.properties.previous;
        } else {
          this.disabledPrevious = true;
        }
```

malep2007 commented:
Test coverage also dropped by over 1% Consider writing more tests @Genza999
